blog/content/post/2020-12-14-trenchboot_attestation.md

```diff
@@ -0,0 +1,294 @@
+
+The communication between attester and verifier is provided by libcoap library.
+It is the C implementation of the Constrained Application Protocol.
+Currently, the communication between the attester and verifier is constrained by
+a maximum transmission unit (~1500 bytes). The block-wise transmission is in the
+development plans. Though we need to fit the public key with the TPM signature
+quote and attestation data in the single transmission package. The following
+snippet adds the attester public key to the attester response.
+
+```C
+	UsefulBufC Tpm2KeyPublic = {attestation_response->tpm2_public_key, attestation_response->tpm2_public_key_len};
+	QCBOREncode_AddBytes(&EC, Tpm2KeyPublic);
+```
+
+Following code allows the verifier to unmarshal the public key from the attestation
+response.
+
+```C
+	/* parse "tpm2_public_key (bytes)" */
+	if((cborerr = charra_cbor_getnext(&DC, &item, QCBOR_TYPE_BYTE_STRING)))
+		goto cbor_parse_error;
+	attestation_response->tpm2_public_key_len = item.val.string.len;
+	attestation_response->tpm2_public_key = (TPM2B_PUBLIC*)item.val.string.ptr;
+```
+
+The verifier must not create its key. Once it will receive the attester
+public key, it must set the handle in its TPM context.
+
+```C
+CHARRA_RC charra_load_external_public_key(ESYS_CONTEXT* ctx,
+	TPM2B_PUBLIC* external_public_key, ESYS_TR* key_handle) {
+	TSS2_RC r = TSS2_RC_SUCCESS;
+	if (external_public_key == NULL) {
+		charra_log_error("External public key do not exist.");
+		return CHARRA_RC_ERROR;
+	}
+
+	r = Esys_LoadExternal(ctx, ESYS_TR_NONE, ESYS_TR_NONE, ESYS_TR_NONE,
+			NULL, external_public_key, TPM2_RH_OWNER, key_handle);
+	if (r != TSS2_RC_SUCCESS ) {
+		charra_log_error("Loading external public key failed.");
+		return CHARRA_RC_ERROR;
+	}
+
+	return CHARRA_RC_SUCCESS;
+}
+```
+
+The following function allows to load external key and it creates `key_handle` that
+is used for the TPM signature verification.
+
+The full scope of changes is available in the following
+[pull request](https://github.com/Fraunhofer-SIT/charra/pull/16).
+
+### Using CHARRA with TPM device
+
+The PoC CHARRA implementation is using the TPM simulator provided in the docker
+container. To use CHARRA with the physical TPM we have created the Yocto layer
+- [meta-trenchboot-attestation](https://github.com/3mdeb/meta-trenchboot-attestation),
+which provides the required libraries, that are used by attester and verifier.
+The Yocto Project (YP) is an open source collaboration project that helps
+developers create custom Linux-based systems regardless of the hardware
+architecture. [meta-trenchboot](https://github.com/3mdeb/meta-trenchboot)
+builds the image which contains the TrenchBoot utilities.
+
+### Proof of concept
+
+To proof the following concept we used the
+[ASRock 4x4 Box R1000V](https://www.asrockind.com/en-gb/4X4%20BOX-R1000V)
+with physical TPM as the attester and the PC with TPM simulator as the verifier.
```

Comment:
Maybe add a reminder that fTPM won't work?

It would be good to finally get a public statement from AMD...

Reply:
I've added the information.